src/Messages.ts

```diff
@@ -0,0 +1,47 @@
+"use strict";
+
+import { panic } from "./Support";
+
+export enum ErrorMessage {
+    ArrayIsInvalidJSONSchema = "An array is not a valid JSON Schema",
+    NullIsInvalidJSONSchema = "null is not a valid JSON Schema",
+    RefMustBeString = "$ref must be a string",
+    AdditionalTypesForbidRequired = "Can't have non-specified required properties but forbidden additionalTypes"
+}
+
+export enum ErrorMessageWithInfo {
+    RefWithFragmentNotAllowed = "Ref URI with fragment is not allowed: %s",
+    InvalidJSONSchemaType = "Value of type %s is not valid JSON Schemas",
+    RequiredMustBeStringOrStringArray = "`required` must be string or array of strings, but is %s",
+    RequiredElementMustBeString = "`required` must only contain strings, but it has %s",
+    TypeMustBeStringOrStringArray = "`type` must be string or array of strings, but is %s",
+    TypeElementMustBeString = "`type` must only contain strings, but it has %s"
+}
+
+export enum ErrorMessageWithTwoInfos {
+    SetOperationCasesIsNotArray = "%s cases must be an array, but is %s"
+}
+
+export function messageError(message: ErrorMessage): never;
+export function messageError(message: ErrorMessageWithInfo, info: string): never;
+export function messageError(message: ErrorMessageWithTwoInfos, info1: string, info2: string): never;
+export function messageError(message: string, ...infos: string[]): never {
+    for (const info of infos) {
+        // FIXME: this won't work if one of the infos, apart from the last, contains "%s"
+        message = message.replace("%s", info);
+    }
+    return panic(message);
```

Comment:
It doesn't look like I will be able to compare these messages to the enum cases and present custom UI based on that. I would prefer that the enum cases are recoverable from the caught error.

Reply:
Yes, I didn't implement the actual exception type yet, I focused on the API for reporting first.

We can use the enum reverse mapping to get to the name of enum case:

```typescript
enum Enum {
    A
}
let a = Enum.A;
let nameOfA = Enum[a]; // "A"
```

Reply:
We could have an exception class that has

* The name of the enum case.
* The "raw" message, with the formatting characters.
* The values to substitute.
* A method that returns the full substituted message.

Would that be enough?

Reply:
Yes that's great.

Reply:
The error should also contain an optional location.